### PR TITLE
Youtube playlist

### PR DESCRIPTION
**Pull request by stellakunzang**
Admin can add youtube playlist, which creates a new tutorial(collection) and adds each individual video to the database. 83.33% coverage currently.

### CODE REVIEW

File: app/controllers/admin/tutorials_controller.rb [hunk 1]
```diff
@@ -3,7 +3,18 @@ def edit
     @tutorial = Tutorial.find(params[:id])
   end
 
-  def create; end
+  def create
+    if results.snippet.nil?
```
Comment (stellakunzang) [inline app/controllers/admin/tutorials_controller.rb:7]: checks that the playlist id is valid

File: app/controllers/admin/tutorials_controller.rb [hunk 2]
```diff
@@ -3,7 +3,18 @@ def edit
...
+      flash[:error] = 'Sorry, that ID is not valid. Try again?'
+      redirect_to new_admin_youtube_playlist_path
+    else
+      tutorial_from_playlist = Tutorial.create(new_tutorial_params)
+      tutorial_from_playlist.create_playlist_videos
+      context = view_context.link_to('View it here', tutorial_path(tutorial_from_playlist.id))
```
Comment (stellakunzang) [inline app/controllers/admin/tutorials_controller.rb:13]: saved to a variable in order to shorten line (rubocop warning)

File: app/controllers/admin/tutorials_controller.rb [hunk 3]
```diff
@@ -28,4 +39,12 @@ def destroy
...
     params.require(:tutorial).permit(:tag_list)
   end
+
+  def results
+    YoutubePlaylistResults.new(params[:playlist_id])
+  end
+
+  def new_tutorial_params
+    results.parameters
+  end
```
Comment (stellakunzang) [inline app/controllers/admin/tutorials_controller.rb:49]: separated these to help with validating playlist_id

File: app/models/tutorial.rb [hunk 4]
```diff
@@ -2,4 +2,28 @@ class Tutorial < ApplicationRecord
...
+      video_params = {}
+      video_params[:title] = video[:snippet][:title]
+      video_params[:description] = video[:snippet][:description]
+      video_params[:video_id] = video[:contentDetails][:videoId]
+      video_params[:thumbnail] = video[:snippet][:thumbnails][:default][:url]
+      video_params[:position] = video[:snippet][:position]
+      playlist_video_params << video_params
+    end
+    playlist_video_params
+  end
```
Comment (stellakunzang) [inline app/models/tutorial.rb:28]: Rubocop said this method was too long, but I couldn't find a quick way to condense it

File: app/models/video.rb [hunk 5]
```diff
@@ -2,4 +2,6 @@ class Video < ApplicationRecord
   has_many :user_videos, dependent: :destroy
   has_many :users, through: :user_videos
   belongs_to :tutorial
+
+  validates :tutorial_id, presence: true
```
Comment (stellakunzang) [inline app/models/video.rb:6]: piece of cake!

File: app/poros/youtube_playlist_results.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    parameters = {}
+    parameters[:title] = snippet[:title]
+    parameters[:description] = snippet[:description]
+    parameters[:thumbnail] = snippet[:thumbnails][:default][:url]
+    parameters[:playlist_id] = @id
+    parameters
+  end
+
+  def snippet
+    return nil if playlist[:items].blank?
```
Comment (stellakunzang) [inline app/poros/youtube_playlist_results.rb:16]: this is where we are actually determining if the playlist id is valid

File: app/services/youtube_service.rb [hunk 7]
```diff
@@ -18,4 +28,8 @@ def conn
       f.params[:key] = ENV['YOUTUBE_API_KEY']
     end
   end
+
+  def part_details
+    'snippet,contentDetails'
+  end
```
Comment (stellakunzang) [inline app/services/youtube_service.rb:34]: created this to shorten line for rubocop, and it was repeated twice so why not?

File: spec/services/youtube_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      expect(playlist_info).to be_an Hash
+      expect(playlist_info).to have_key :items
+    end
+
+    it "playlist_items_info" do
+      playlist_items_info = @service.playlist_items_info(@id)
+      expect(playlist_items_info[:items].length).to eq(13)
+    end
+
+    xit "can handle more than 50" do
```
Comment (stellakunzang) [inline spec/services/youtube_service_spec.rb:21]: Left this unfinished, but we've got the placeholder to remind us

**Comment by stellakunzang** [comment]
Travis worked on the first try!

File: app/controllers/admin/tutorials_controller.rb (hunk 1 above)
Comment (perryr16) [reply app/controllers/admin/tutorials_controller.rb:7]: excellent

File: app/models/tutorial.rb (hunk 4 above)
Comment (perryr16) [reply app/models/tutorial.rb:28]: Agreed, I'm trying to research this

File: spec/features/admin/admin_can_import_playlist_spec.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    visit new_admin_tutorial_path
+    click_on "Import YouTube Playlist"
+
+    expect(current_path).to eq(new_admin_youtube_playlist_path)
+
+    fill_in :playlist_id, with: "12345abcde"
+    click_on "Import Playlist"
+
+    expect(current_path).to eq(new_admin_youtube_playlist_path)
+    expect(page).to have_content("Sorry, that ID is not valid. Try again?")
```
Comment (perryr16) [inline spec/features/admin/admin_can_import_playlist_spec.rb:45]: Excellent sad path testing

**Review by perryr16 — COMMENTED** [review]
Looks great. Appreciate the sad path/validity work

**Pull request merged**